Fix: Add more space to bottom docked button

truph01 opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
---> 
$ https://github.com/Expensify/App/issues/44056
PROPOSAL: https://github.com/Expensify/App/issues/44056#issuecomment-2179772696


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
1. Open app
2. Tap avatar
4. Tap display name
5. Observe the space at the bottom of docked button
6. There should be proper bottom SafeSpace below the button
- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->
<img width="361" alt="Screenshot 2024-07-03 at 14 10 53" src="https://github.com/Expensify/App/assets/170341821/3407aace-1a60-478f-9fb0-f99f0713371b">

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->
<img width="347" alt="Screenshot 2024-07-03 at 14 22 15" src="https://github.com/Expensify/App/assets/170341821/8dd0d8e3-7600-4fce-a039-e42c01211275">

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->
<img width="375" alt="Screenshot 2024-07-03 at 14 09 59" src="https://github.com/Expensify/App/assets/170341821/9290eb40-ebe5-43e0-9310-43d6c2eda062">

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->
<img width="369" alt="Screenshot 2024-07-03 at 14 21 09" src="https://github.com/Expensify/App/assets/170341821/11fa8a21-0cee-412e-b487-5ee539acd3a6">

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->
<img width="1512" alt="Screenshot 2024-07-03 at 14 19 09" src="https://github.com/Expensify/App/assets/170341821/e19588cd-0475-45fc-8d99-d15b29cfa95c">

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->
<img width="1196" alt="Screenshot 2024-07-03 at 14 17 58" src="https://github.com/Expensify/App/assets/170341821/12d5ff49-f635-4774-b800-10dfaccb3b1f">

</details>

truph01 committed:
Fix: Add more space to bottom docked button

shawnborton commented:
Looking good, but here are spots where it's missing:
- Wallet > Transfer Balance page
- Workspace settings > Members > Invite new members

On the Edit expense details > Date screen, we have too much space below the Save button:
![CleanShot 2024-07-03 at 10 16 13@2x](https://github.com/Expensify/App/assets/2319350/74823ea4-c43e-4233-8bb0-f1e10f73d768)


One interesting observation though... I feel like when the keyboard is up, we don't need the SafeSpace to show. Is there something we can do about that? Otherwise it feels like the gap below the Save button is way too large when the keyboard is visible. 
![CleanShot 2024-07-03 at 10 16 01@2x](https://github.com/Expensify/App/assets/2319350/09c99b5e-1e32-420c-9996-4af06d2fb5d5)

truph01 committed:
Fix: IOU request step screen

truph01 committed:
Fix: Remove additional padding bottom if keyboard is shown

truph01 commented:
@shawnborton 
> Looking good, but here are spots where it's missing:
Wallet > Transfer Balance page
Workspace settings > Members > Invite new members

- These screens do not use a form wrapper component, but the fix in this PR just applies to the bottom button in form wrapper component. If we need to fix these screen, we need to add the padding bottom to its bottom button manually. WDYT?

> On the Edit expense details > Date screen, we have too much space below the Save button:

- I fixed it

> One interesting observation though... I feel like when the keyboard is up, we don't need the SafeSpace to show. Is there something we can do about that? Otherwise it feels like the gap below the Save button is way too large when the keyboard is visible.

- I fixed it. Below is the result:

https://github.com/Expensify/App/assets/170341821/590d8171-4c52-41f0-8ee3-aeb183fb155f

truph01 committed:
Fix: Lint

shawnborton commented:
> we need to add the padding bottom to its bottom button manually. WDYT?

I think we should add the padding there for consistency. And we should check as many other pages as we can too to make sure we aren't missing any. 

> I fixed it. Below is the result:

Much better!

truph01 committed:
Fix: Create useExtraSafePaddingBottomStyle

truph01 committed:
Fix: Merge main

truph01 committed:
Fix: extraPaddingBottomStyle logic

truph01 committed:
Fix: Move extra padding bottom to FormAlertWithSubmitButton

truph01 committed:
Fix: Remove unsed changes

truph01 commented:
@allgandalf I updated PR. Now, I use the additional padding bottom style in ```FormAlertWithSubmitButton``` directly instead of styling ```<FormAlertWithSubmitButton containerStyles={[..., additionalPaddingBottomStyle]} ... />``` in ```FormWrapper``` as I did before because, there is a lot of screens we do not use ```FormWrapper``` but still have docked button like mentioned above:

> Looking good, but here are spots where it's missing:
Wallet > Transfer Balance page
Workspace settings > Members > Invite new members

truph01 committed:
Fix: Lint

truph01 commented on src/hooks/useExtraSafePaddingBottomStyle.ts:
```diff
@@ -0,0 +1,34 @@
+const useExtraSafePaddingBottomStyle = () => {
+    const styles = useThemeStyles();
+    const [willKeyboardShow, setWillKeyboardShow] = useState(false);
+    useEffect(() => {
+        const keyboardWillShowListener = Keyboard.addListener('keyboardWillShow', () => {
+            setWillKeyboardShow(true);
+        });
```
Ah, I missed your question. That ```useEffect``` will listen to the keyboard state to know when the keyboard will show or hide.

allgandalf reviewed (commented):
also are we covering all the cases with these changes?

truph01 commented on src/hooks/useExtraSafePaddingBottomStyle.ts:
```diff
@@ -0,0 +1,34 @@
+
+const useExtraSafePaddingBottomStyle = () => {
+    const styles = useThemeStyles();
+    const [willKeyboardShow, setWillKeyboardShow] = useState(false);
```
- I don't think we should we the function you mentioned, because it is not used in our app, just in test file.
- We already have the way to get the keyboard state, see this [file](https://github.com/Expensify/App/blob/main/src/components/withKeyboardState.tsx). But it use ```keyboardDidShow``` and ```keyboardDidHide``` events to check if the keyboard is shown or not. It works well but there is an issue: when the keyboard is shown (user can see the keyboard), the bottom button is still applied the ```padding-bottom: 20px``` for a while. So I use ```keyboardWillShow``` and ```keyboardWillHide``` to fix that issue.

truph01 commented:
> also are we covering all the cases with these changes?

Yes

truph01 committed:
Feat: Merge maiin

truph01 committed:
Feat: Add typeto state

truph01 commented on src/hooks/useExtraSafePaddingBottomStyle.ts:
```diff
@@ -0,0 +1,34 @@
+
+const useExtraSafePaddingBottomStyle = () => {
+    const styles = useThemeStyles();
+    const [willKeyboardShow, setWillKeyboardShow] = useState(false);
```
I added the type for it. But I don't think we should rename the state, because as I mentioned above, ```willKeyboardShow``` is used to fix the issue ```didKeyboardShow``` encountered: There is a delay when the keyboard is visible by the user and the when ```didKeyboardShow``` is set to ```true```. So it is not just ```isKeyboardVisible```.

truph01 commented:
@allgandalf Plz help review PR again. Thanks

shawnborton commented:
Is this one ready for testing again? Would love to get this one over the finish line!

shawnborton commented:
Design-wise this is looking good, but let's still run it through a proper C+ testing. 

Can you try pulling main though so I can rebuild the test builds? From the test build above, a lot of the page transitions and such feel a bit jumpy or not very smooth... so just want to make sure this PR isn't causing any of that. Let me know if you are experiencing something similar.

truph01 committed:
Merge branch 'Expensify:main' into fix/44056

shawnborton commented:
Cool, will rebuild the test builds and see how it feels. I think we can move forward with the checklist and final review though in the meantime!

shawnborton commented:
Okay nice, this feels way better. I say we get this into final review!

shawnborton commented:
Hmm how are we calculating that space? How does the space below other buttons on this same platform look?

shawnborton commented:
Yeah, that looks better. Can you help me understand why the two of them are different though?

allgandalf reviewed (approved):
changes LGTM, looks good to be merged into staging

shawnborton commented:
Got it. Yeah, ideally we use whatever is consistent with other places (like display name). How do you recommend we fix that?

shawnborton commented:
Sounds good!

puneetlath commented on src/hooks/useExtraSafePaddingBottomStyle.ts:
```diff
@@ -0,0 +1,34 @@
+import useStyledSafeAreaInsets from './useStyledSafeAreaInsets';
+import useThemeStyles from './useThemeStyles';
+
+const useExtraSafePaddingBottomStyle = () => {
```
Perhaps we can name it something that makes it more clear when it can be used. "extra safe" is a bit ambiguous.

puneetlath commented:
@Expensify/design someone mind reviewing as well?

puneetlath commented:
Oh whoops, my bad. I see @shawnborton already reviewed above.

truph01 committed:
merge main

truph01 committed:
Feat: Rename useExtraSafePaddingBottomStyle to useSafePaddingBottomStyle

truph01 commented:
@puneetlath I addressed your comment above

puneetlath commented on src/hooks/useSafePaddingBottomStyle.ts:
```diff
@@ -0,0 +1,34 @@
+import useStyledSafeAreaInsets from './useStyledSafeAreaInsets';
+import useThemeStyles from './useThemeStyles';
+
+const useSafePaddingBottomStyle = () => {
```
Can we add some general documentation above about what this hook is for.

puneetlath replied:
Sorry for being a stickler, but I'm finding this comment a bit hard to follow. In the future, when should someone use this hook? Maybe if you can just explain that to me in words in a comment here, I can help you craft a code comment that is a bit more clear.

truph01 replied:
@puneetlath

> In the future, when should someone use this hook

This hook, useSafePaddingBottomStyle, is useful for adding extra bottom padding to a component based on the device's safe area. If the device's safe area padding bottom is 0, the hook returns 0. Otherwise, it provides a padding bottom of 20. Use this to ensure content visibility and layout consistency across different devices.

puneetlath replied:
That's beautiful. Very clear! Let's make that the comment.

puneetlath commented on src/components/FormAlertWithSubmitButton.tsx:
```diff
@@ -79,10 +80,11 @@ function FormAlertWithSubmitButton({
 }: FormAlertWithSubmitButtonProps) {
     const styles = useThemeStyles();
     const style = [!footerContent ? {} : styles.mb3, buttonStyles];
+    const extraSafePaddingBottomStyle = useSafePaddingBottomStyle();
```
To match the hook name. 
```suggestion
    const safePaddingBottomStyle = useSafePaddingBottomStyle();
```

puneetlath reviewed (changes requested):
Sorry a couple more small requests.

truph01 committed:
Feat: Add comment to hooks and rename variable

truph01 commented:
All comment are resolved

puneetlath commented on src/hooks/useSafePaddingBottomStyle.ts:
```diff
@@ -0,0 +1,37 @@
+import useStyledSafeAreaInsets from './useStyledSafeAreaInsets';
+import useThemeStyles from './useThemeStyles';
+
+// Refer to https://github.com/Expensify/App/issues/44056, the default padding bottom of any device is not enough.
+// Devices with 0 padding bottom value, this hook returns { paddingBottom: 0 }.
+// Otherwise, it returns { paddingBottom: 20 }.
```
```suggestion
// This hook is useful for adding extra bottom padding to a component based on the device's safe area
// If the device's safe area padding bottom is 0, the hook returns 0. Otherwise, it provides a padding bottom of 20.
// Use this to ensure content visibility and layout consistency across different devices.
```

truph01 committed:
Update comment src/hooks/useSafePaddingBottomStyle.ts

Co-authored-by: Puneet Lath <puneet@expensify.com>

truph01 commented:
I updated PR @puneetlath

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/10105962433) to staging by https://github.com/puneetlath in version: 9.0.13-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/10149765247) to production by https://github.com/luacmartins in version: 9.0.13-4 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|failure ❌
